src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,109 @@
+use alloc::vec::Vec;
+use core::iter;
+
+use super::{EmptySubtreeRoots, MerklePath, RpoDigest};
+
+/// A different representation of [`MerklePath`] designed for memory efficiency for Merkle paths
+/// with empty nodes.
+///
+/// Empty nodes in the path are stored only as their position, represented with a bitmask. A
+/// maximum of 64 nodes in the path can be empty. The number of empty nodes has no effect on memory
+/// usage by this struct, but will incur overhead during iteration or conversion to a
+/// [`MerklePath`], for each empty node.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct SparseMerklePath<const DEPTH: u8> {
+    /// A bitmask representing empty nodes. The set bit corresponds to the depth of an empty node.
+    empty_nodes: u64,
+    /// The non-empty nodes, stored in depth-order, but not contiguous across depth.
+    nodes: Vec<RpoDigest>,
+}
+
+impl<const DEPTH: u8> SparseMerklePath<DEPTH> {
+    /// Converts a Merkle path to a sparse representation.
+    pub fn from_path(path: MerklePath) -> Self {
+        // Note that the path does not include the node itself that it is a path to.
+        // That is to say, the path is not inclusive of its ending.
+        assert!(path.len() <= u8::MAX.into(), "SparseMerklePath may have at most 256 items");
```

Comment:
We should check against `SMT_MAX_DEPTH` here rather than `u8::MAX`. Also, I wonder if we should return an error here instead of panicking.

Reply:
I made it an assert to match https://github.com/0xPolygonMiden/crypto/blob/aa45474377e978050958958d75688e7a8d46b628/miden-crypto/src/merkle/path.rs#L26

should I change both of them, or just this one?